x/gamm/keeper/multihop_test.go

```diff
@@ -55,74 +92,92 @@ func (suite *KeeperTestSuite) TestBalancerPoolSimpleMultihopSwapExactAmountIn()
-				SwapFee: poolDefaultSwapFee,
-				ExitFee: sdk.NewDec(0),
-			})
 
-			// if we expect a reduced fee to apply, we set both pools in DistrInfo to replicate it being an incentivized pool
-			// each pool has three gauges, hence 6 gauges for 2 pools
-			if test.reducedFeeApplied {
-				test := poolincentivestypes.DistrInfo{
-					TotalWeight: sdk.NewInt(6),
-					Records: []poolincentivestypes.DistrRecord{
-						{GaugeId: 1, Weight: sdk.OneInt()}, {GaugeId: 2, Weight: sdk.OneInt()}, {GaugeId: 3, Weight: sdk.OneInt()},
-						{GaugeId: 4, Weight: sdk.OneInt()}, {GaugeId: 5, Weight: sdk.OneInt()}, {GaugeId: 6, Weight: sdk.OneInt()}},
-				}
-				suite.App.PoolIncentivesKeeper.SetDistrInfo(suite.Ctx, test)
+			// Create pools desired swap fee, fund acc with respective coins
+			for i, coin := range test.param.poolAssets {
+				suite.PrepareBalancerPoolWithCoins(coin[0], coin[1])
+				suite.FundAcc(suite.TestAccs[0], sdk.NewCoins(coin[0], coin[1]))
+				suite.updatePoolSwapFee(suite.Ctx, uint64(i+1), test.param.poolFee[i])
 			}
 
-			// Calculate the chained spot price.
-			calcSpotPrice := func() sdk.Dec {
-				dec := sdk.NewDec(1)
-				tokenInDenom := test.param.tokenIn.Denom
-				for i, route := range test.param.routes {
-					if i != 0 {
-						tokenInDenom = test.param.routes[i-1].TokenOutDenom
-					}
-					pool, err := keeper.GetPoolAndPoke(suite.Ctx, route.PoolId)
-					suite.NoError(err, "test: %v", test.name)
-
-					sp, err := pool.SpotPrice(suite.Ctx, tokenInDenom, route.TokenOutDenom)
-					suite.NoError(err, "test: %v", test.name)
-					dec = dec.Mul(sp)
+			// if test specifies incentivized gauges, set them here
+			if len(test.param.incentivizedGauges) > 0 {
+				var records []poolincentivestypes.DistrRecord
+				totalWeight := sdk.NewInt(int64(len(test.param.incentivizedGauges)))
+				for _, gauge := range test.param.incentivizedGauges {
+					records = append(records, poolincentivestypes.DistrRecord{GaugeId: gauge, Weight: sdk.OneInt()})
 				}
-				return dec
+				distInfo := poolincentivestypes.DistrInfo{
+					TotalWeight: totalWeight,
+					Records:     records,
+				}
+				suite.App.PoolIncentivesKeeper.SetDistrInfo(suite.Ctx, distInfo)
 			}
 
-			calcOutAmountAsSeparateSwaps := func(adjustedPoolSwapFee sdk.Dec) sdk.Coin {
+			// calcOutAmountAsSeparateSwaps calculates the multi-hop swap as separate swaps, while also
+			// utilizing a cacheContext so the state does not change
+			calcOutAmountAsSeparateSwaps := func(osmoFeeReduced bool) sdk.Coin {
 				cacheCtx, _ := suite.Ctx.CacheContext()
 
-				if adjustedPoolSwapFee != poolDefaultSwapFee {
+				if osmoFeeReduced {
 					for _, hop := range test.param.routes {
-						err := suite.updatePoolSwapFee(cacheCtx, hop.PoolId, adjustedPoolSwapFee)
-						suite.NoError(err, "test: %v", test.name)
+						// determine the expected reduced swap fee
+						hopPool, _ := keeper.GetPoolAndPoke(cacheCtx, hop.PoolId)
+						currentPoolSwapFee := hopPool.GetSwapFee(cacheCtx)
```

Comment:
would it be possible to add inline comments on what is happening here? It was kind of hard for me to follow why we need to update the pool swap fees directly in the beginning

Reply:
From my understanding, I think it's getting / calculating the swap fees manually and then updating the pool swap fee manually in accordance to the calculated swap fee, and then compare it with the calculation that has been done via multihop swap method.